client/statement-store/src/lib.rs

```diff
@@ -0,0 +1,986 @@
+	}
+}
+
+/// Statement store.
+pub struct Store {
+	db: parity_db::Db,
+	index: RwLock<Index>,
+	validate_fn: Box<
+		dyn Fn(
+				Option<BlockHash>,
+				StatementSource,
+				Statement,
+			) -> std::result::Result<ValidStatement, InvalidStatement>
+			+ Send
+			+ Sync,
+	>,
+	time_override: Option<u64>,
+	metrics: PrometheusMetrics,
+}
+
+#[derive(Encode, Decode, Clone)]
+struct StatementMeta {
+	priority: u64,
+	timestamp: u64,
+}
+
+#[derive(Encode, Decode)]
+struct StatementWithMeta {
+	meta: StatementMeta,
+	statement: Statement,
+}
+
+enum IndexQuery {
+	Unknown,
+	Exists(u64),
+	Expired(u64),
+}
+
+impl Index {
+	fn insert_with_meta(&mut self, hash: Hash, statement: StatementWithMeta) {
+		let mut all_topics = [None; 4];
+		let mut nt = 0;
+		let StatementWithMeta { statement, meta } = statement;
+		while let Some(t) = statement.topic(nt) {
+			self.by_topic.entry(t).or_default().insert(hash);
+			all_topics[nt] = Some(t);
+			nt += 1;
+		}
+		let key = statement.decryption_key();
+		if let Some(k) = &key {
+			self.by_dec_key.entry(k.clone()).or_default().insert(hash);
+		}
+		if nt > 0 || key.is_some() {
+			self.all_topics.insert(hash, (all_topics, key));
+		}
+		self.expired.remove(&hash);
+		if self.entries.insert(hash, meta.clone()).is_none() {
+			self.by_priority.push(EvictionPriority {
+				hash,
+				priority: meta.priority,
+				timestamp: meta.timestamp,
+			});
+		}
+	}
+
+	fn query(&self, hash: &Hash) -> IndexQuery {
+		if let Some(meta) = self.entries.get(hash) {
+			return IndexQuery::Exists(meta.priority)
+		}
+		if let Some(meta) = self.expired.get(hash) {
+			return IndexQuery::Expired(meta.priority)
+		}
+		IndexQuery::Unknown
+	}
+
+	fn insert_expired(&mut self, hash: Hash, meta: StatementMeta) {
+		self.expired.insert(hash, meta);
+	}
+
+	fn is_expired(&self, hash: &Hash) -> bool {
+		self.expired.contains_key(hash)
+	}
+
+	fn iter(
+		&self,
+		key: Option<DecryptionKey>,
+		topics: &[Topic],
+		mut f: impl FnMut(&Hash) -> Result<()>,
+	) -> Result<()> {
+		let mut sets: [Option<&HashSet<Hash>>; 4] = Default::default();
+		let mut num_sets = 0;
+		for t in topics {
+			sets[num_sets] = self.by_topic.get(t);
+			if sets[num_sets].is_some() {
+				num_sets += 1;
+			}
+		}
+		if num_sets == 0 && key.is_none() {
+			// Iterate all entries
+			for h in self.entries.keys() {
+				log::trace!(target: LOG_TARGET, "Iterating: {:?}", HexDisplay::from(h));
+				f(h)?
+			}
+		} else {
+			// Start with the smallest topic set or the key set.
+			sets[0..num_sets].sort_by_key(|s| s.map_or(0, HashSet::len));
+			if let Some(key) = key {
+				let key_set =
+					if let Some(set) = self.by_dec_key.get(&key) { set } else { return Ok(()) };
+				for item in key_set {
+					if sets.iter().all(|set| set.unwrap().contains(item)) {
+						log::trace!(
+							target: LOG_TARGET,
+							"Iterating by key: {:?}",
+							HexDisplay::from(item)
+						);
+						f(item)?
+					}
+				}
+			} else {
+				for item in sets[0].unwrap() {
```

Comment:
we never iterate over statements that mathes 1 topic, even of 1 topic was passed by user